fenix/app/src/main/java/org/mozilla/fenix/components/appstate/AppAction.kt

```diff
@@ -30,6 +31,28 @@ import org.mozilla.fenix.wallpapers.Wallpaper
  * [Action] implementation related to [AppStore].
  */
 sealed class AppAction : Action {
+    /**
+     * [AppAction] dispatched to indicate that the store is initialized and
+     * ready to use. This action is dispatched automatically before any other
+     * action is processed. Its main purpose is to trigger initialization logic
+     * in middlewares. The action itself should have no effect on the [AppState].
+     */
+    object Init : AppAction()
+
+    /**
+     * NOTE: This action is not yet functional and will require https://bugzilla.mozilla.org/show_bug.cgi?id=1845409
+     * to be resolved. This is part of an ongoing lib-state refactor.
```

Comment:
Seeing TODO type comments usually gives me a bit of a pause, couple thoughts here are that we may have grown this PR a bit too big in terms of refactoring considering some of the changes are not yet functional and also seeing the additional wallpaper reducer refactoring. It would probably be better to not land non-functional changes because (1) they are only half-functional in this case and increases the maintenance burden if they were fully non-functional, and (2) there's no guarantee we will get to reference bug any time soon and can cause confusion.

Now I don't really know for sure if this is actually non-functional given the fact that I see this in the reducer, but I am assuming we actually want to open a tab instead of just changing the mode. 

I would encourage future work to try and avoid doing too much in a big PR.

Reply:
I do generally agree with all of the above of course. You are right that the action is not entirely non-functional, since it's used in the reducer. I can update the comment accordingly if you'd like. 

But yes, I would not consider this fully functional because it does not actually trigger the logic to open a tab and navigate to the browser (yet). The unfortunate truth is that much of this refactoring work is deeply inter-related, so I'm not sure there is an ordering of work that will let us avoid temporary hackiness like this. I actually started on the `BrowsingMode` stuff because it was used in `HomeActivity::openToBrowserAndLoad` and I thought refactoring that method would be easier if this state already lived in the Store.

Finally, on the size of the PR. I do completely agree, which is why this patch was initially split into 4 commits with the offer to split into multiple patches if reviewers desired. However, I do think that small patches are intended to accomplish the following:

1. reduce risk of unintended breakage
2. allow simple backouts
3. ease review burden

For the first, I think one benefit by landing something of this size is that it allows QA to do a thorough smokepass of the feature entirely, instead of expecting them to do so regularly as individual patches are landed.

For the second, I do think that lots of small patches would quickly become difficult to revert individually, because I think there is risk of small follow-up updates being included in later patches. That can be attributed to engineer error, but it's still a real risk. 

For the last, @boek and I had come to some agreements to just go ahead and do the thing while we were here.